s3.go

```diff
@@ -56,6 +56,22 @@ type Config struct {
 	RootDirectory       string
 	Workers             int
 	CredentialsEndpoint string
+	KeyTransform        string
+}
+
+var KeyTransforms = map[string]func(ds.Key) string{
+	"default": func(k ds.Key) string {
+		return k.String()
+	},
+	"suffix": func(k ds.Key) string {
+		return k.String() + "/data"
+	},
+	"next-to-last/2": func(k ds.Key) string {
+		s := k.String()
+		offset := 1
+		start := len(s) - 2 - offset
+		return s[start:start+2] + "/" + s
+	},
```

Comment:
I think there are some problems here

```
datastore.NewKey("a") -> panic
datastore.NewKey("ab") -> /a//ab
datastore.NewKey("abc") -> ab//abc
datastore.NewKey("abcd") -> bc//abcd
```

* Panics on keys w/ len == 1, should return an error instead
* Keys w/ len == 2 have a slash in their prefix (should probably return an error instead?)
* Keys w/ len > 2 have two slashes instead of one

A small unit test would be nice here.

Reply:
Thanks for this @guseggert ! In practice, I am not seeing the double slash issue. Also just running through the logic I don't see how that would be possible but I'm new to Go 🙈.

However for keys of len 2 or smaller I agree this should be handled differently.
For keys of length 3 or more it still seems correct given a param of 2.

Reply:
All I did was add this test to print out the results, should help you reproduce:

```go
func TestTransforms(t *testing.T) {
	transform := KeyTransforms["next-to-last/2"]
	keys := []string{"ab", "abc", "abcd"}
	for _, k := range keys {
		fmt.Printf("datastore.NewKey(\"%s\") -> %s\n", k, transform(datastore.NewKey(k)))
	}
}
```


Full test:


```go
func TestTransforms(t *testing.T) {
	cases := []struct {
		transform     string
		key           string
		expectedS3Key string
	}{
		{
			transform:     "next-to-last/2",
			key:           "ab",
			expectedS3Key: "ab/ab",
		},
		{
			transform:     "next-to-last/2",
			key:           "abc",
			expectedS3Key: "ab/abc",
		},
		{
			transform:     "next-to-last/2",
			key:           "abcd",
			expectedS3Key: "ab/abcd",
		},
	}
	for _, c := range cases {
		t.Run(fmt.Sprintf("%s %s", c.transform, c.key), func(t *testing.T) {
			transform := KeyTransforms[c.transform]
			dsKey := datastore.NewKey(c.key)
			s3Key := transform(dsKey)
			if s3Key != c.expectedS3Key {
				t.Fatalf("expected %s, got %s", c.expectedS3Key, s3Key)
			}
		})
	}
}
```

output:

```
go test -v -run TestTransforms\$ . 
=== RUN   TestTransforms
=== RUN   TestTransforms/next-to-last/2_ab
    s3_test.go:105: expected ab/ab, got /a//ab
=== RUN   TestTransforms/next-to-last/2_abc
    s3_test.go:105: expected ab/abc, got ab//abc
=== RUN   TestTransforms/next-to-last/2_abcd
    s3_test.go:105: expected ab/abcd, got bc//abcd
--- FAIL: TestTransforms (0.00s)
    --- FAIL: TestTransforms/next-to-last/2_ab (0.00s)
    --- FAIL: TestTransforms/next-to-last/2_abc (0.00s)
    --- FAIL: TestTransforms/next-to-last/2_abcd (0.00s)
FAIL
FAIL	github.com/3box/go-ds-s3	0.004s
FAIL
```